tests/test_mnemonic_generation.py

```diff
@@ -195,3 +218,95 @@ def test_50_dice_rolls():
+
+    mnemonic = mnemonic_generation.generate_mnemonic_from_coin_flips(coin_flips)
+    actual = " ".join(mnemonic)
+    assert bip39.mnemonic_is_valid(actual)
+    assert actual == expected
+
+    coin_flips = "1111000101001010101010101010010000100111100000101010000101101101111000000101110101010010101010010101010001001011010101111101001001000010111110101011111101010010011010101001010111010100000011111001010010101101001010010101010010101010100101010010100101001001"
+    expected = "vanish fetch poverty excuse claw report lift prevent power pelican sting pig cook garden endless famous staff lake clip famous enjoy enhance pioneer click"
+
+    mnemonic = mnemonic_generation.generate_mnemonic_from_coin_flips(coin_flips)
+    actual = " ".join(mnemonic)
+    assert bip39.mnemonic_is_valid(actual)
+    assert actual == expected
+
+
+
+def test_128_coin_flips():
+    """ 128 coin flips input should yield the same 12-word mnemonic as iancoleman.io/bip39 """
+    # Check "Show entropy details", paste in coin flip sequence, click "Binary", select "Mnemonic Length" as "Use Raw Entropy (3 words per 32 bits)"
+    coin_flips = "10101010101011101100010000000011001000110000000000110000000010011100000000000000100000000001100000011100101100100001001100110010"
+    expected = "primary involve absorb ecology adapt agent abandon avoid blossom tortoise luggage grape"
+    mnemonic = mnemonic_generation.generate_mnemonic_from_coin_flips(coin_flips)
+    actual = " ".join(mnemonic)
+    assert bip39.mnemonic_is_valid(actual)
+    assert actual == expected
+
+    coin_flips = "00000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000"
+    expected = "abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon about"
+    mnemonic = mnemonic_generation.generate_mnemonic_from_coin_flips(coin_flips)
+    actual = " ".join(mnemonic)
+    assert bip39.mnemonic_is_valid(actual)
+    assert actual == expected
+
+    coin_flips = "10100010100101010110101011010001010011101011111010010101010010101010101010010100101010101001010101010110101011111111000001001000"
+    expected = "penalty prize reform output true pipe prevent next next rely winter museum"
+    mnemonic = mnemonic_generation.generate_mnemonic_from_coin_flips(coin_flips)
+    actual = " ".join(mnemonic)
+    assert bip39.mnemonic_is_valid(actual)
+    assert actual == expected
+
+    
+def test_coin_flips_invalid_input():
+    """
+        Should raise a ValueError if coin_flips contains invalid characters or has an invalid length.
+    """
+    # Test invalid characters
+    with pytest.raises(ValueError) as e:
+        invalid_coin_flips = "01012" + "0" * 123  # 128 chars with a '2'
+        mnemonic_generation.generate_mnemonic_from_coin_flips(invalid_coin_flips)
+    assert "only contain '0' or '1'" in str(e)
```

Comment:
This sort of assert might be a bit fragile, though any future refactoring of the error message would trigger an obvious failure here so it would be quickly caught.

I don't have a suggestion for an improvement. Just noting that this style of test always makes me uncomfortable, even when I do it myself.